### DISCUSSION TITLE
z.or() typescript is wrong

### DISCUSSION THREAD

**Discussion by ghostlexly** [colinhacks/zod#2803] [Q&A] [answered]
Hey,

With this code:

```ts
const airportSchema = z
  .object({
    id: z.number().int().positive(),
  })
  .or(airportsStoreSchema.body);

const body = z.object({
  company: z.string().min(2).max(255),
  state: z.nativeEnum(States),
  lta: z.string().min(2).max(255),
  flightNumber: z.string().min(2).max(255),
  departureTime: z.coerce.date(),
  arrivalTime: z.coerce.date(),
  handOverTime: z.coerce.date().optional(),
  expeditor: z.string().min(2).max(255),
  weightLimit: z.number().min(0),
  volumeLimit: z.number().min(0),
  quantityLimit: z.number().min(0),
  departureAirport: airportSchema,
  arrivalAirport: airportSchema,
});
```

I should be able to call `flight.departureAirport.id` and `flight.departureAirport.name` which is inside airportsStoreSchema.body but typescript tells me that .id doesn't exist.

**Comment by willruggiano** [comment]
You can't just call `flight.departureAirport.id` (unless `airportsStoreSchema.body` also has an id field). The type of `airportSchema` is a union, and therefore you need to narrow it down in your code (using `typeof`, `in`, etc). Typescript can't just magically tell you what it is.

So like:
```typescript
if ("id" in flight.departureAirport) {
  // flight.departureAirport is of type { id: number; }
} else {
  // flight.departureAirport is of type z.infer<typeof airportsStoreSchema.body>
}
```

https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#union-types

**Comment by ghostlexly** [comment]
Awesome that worked, thanks for your help @willruggiano .

Perhaps, do you know the inverse of this code ?

i tryed this one but it's not working:

```ts
if (!("id" in body.departureAirport)) {
  // Code to be executed if "id" is not in body.departureAirport
}
```

**Comment by willruggiano** [comment, marked as answer]
You could just do:
```typescript
if ("id" in body.departureAirport) {
  // code to be executed if "id" is in body.departureAirport
} else {
  // code to be executed otherwise
}
```

Also remember that your code might be different depending on how many types exist in the union. So if body.departureAirport is a union of _three_ disparate types:
```typescript
type T =
  | {
      id: string;
    } 
  | {
      body: string;
    }
  | {
      foobar: boolean;
    };
```

then the code: 
```typescript
if ("id" in body.departureAirport) {
  // code to be executed if typeof body.departureAirport === { id: string }
} else {
  // Typescript needs more help here. It just knows you *don't* have a { id: string }
  // but doesn't know which of { body: string } | { foobar: boolean } it is
}
```

needs some adjusting to compensate for your type union.

**Comment by willruggiano** [comment]
Tangentially related, and totally dependent on what you are trying to do, but I have found that _discriminated_ unions are immensely helpful in cases like this.

```typescript
const AirportSchema = z.discriminatedUnion("type", [
  z.object({  // the { id: string } case
    type: z.literal("airport-id"),
    id: z.string()
  }),
  z.object({  // the { body: string } case
    type: z.literal("airport-body"),
    body: z.string()
  }),
  z.object({  // the { foobar: boolean } case
    type: z.literal("airport-foobar"),
    foobar: z.boolean()
  })
]);
type Airport = z.infer<typeof AirportSchema>;
```

then you can do:
```typescript
switch (a.type) {
  case "airport-id":
    a.id; // ok; typeof a === { id: string }
  case "airport-body":
    a.body; // ok; typeof a === { body: string }
  case "airport-foobar":
    a.foobar; // ok; typeof a === { foobar: boolean }
}
```

Again, totally depends what you are trying to do. I have found discriminated unions to be super super powerful though in my personal experience, even outside the context of zod (it's all typescript after all)

**Comment by JacobWeisenburger** [comment]
Was this question answered? I'd like to close this issue if possible.
